### PR TITLE
[RabbitMQ pub/sub] Allow quorum queues

### PR DESCRIPTION
**Pull request by aaguilartablada**
# Description

Allow Dapr user creating quorum queues through Subscription metadata

## Issue reference

https://github.com/dapr/components-contrib/issues/2814
https://github.com/dapr/components-contrib/issues/2544

Please reference the issue this PR will close: #_[2814]_ #_[2544]_

### CODE REVIEW

**Review by berndverst — COMMENTED** [review]
Please add an integration tests in `tests/certification/pubsub/rabbitmq` it should be a new test case there.

**Comment by DeepanshuA** [comment]
A similar PR was raised earlier with multiple queue options, but that seems to have gone stale: https://github.com/dapr/components-contrib/pull/2556

**Comment by berndverst** [comment]
> A similar PR was raised earlier with multiple queue options, but that seems to have gone stale: #2556

Great point @DeepanshuA.

@aaguilartablada please take a look at #2556 and incorporate the feedback we left on that PR. We will want to see those details here as well.

As you are probably the most knowledgeable about RabbitMQ (among maintainers and approvers) @DeepanshuA I'll defer to you for the review of this PR.

Unfortunately this PR here isn't quite ready and we are now approaching code freeze any day.

**Comment by aaguilartablada** [comment]
Sorry for the delay. I was out for some days. I have just uploaded changes to allow (also) stream queues and apply x-max-length, x-max-length-bytes and x-max-age when defined.

To cover these metadata cases in certification tests I have included x-queue-type and x-max-length arguments in the subscription generated to run TestRabbitMQ.

File: pubsub/rabbitmq/rabbitmq.go
Comment (berndverst) [inline pubsub/rabbitmq/rabbitmq.go:446]: If this is not defined at the request level do you want to use the global default you define in the component metadata? `r.rabbitMQMetadata.MaxAge` -- you made that value configurable but don't use it anywhere.

File: pubsub/rabbitmq/rabbitmq.go
Comment (berndverst) [inline pubsub/rabbitmq/rabbitmq.go:410]: If queue type isn't specified at the request level, do you want to use the global component queue type you made configurable? `r.rabbitMQMetadata.QueueType`

File: pubsub/rabbitmq/metadata.go
Comment (berndverst) [inline pubsub/rabbitmq/metadata.go:54]: Currently unused.

File: pubsub/rabbitmq/metadata.go
Comment (berndverst) [inline pubsub/rabbitmq/metadata.go:48]: Currently unused.

File: tests/certification/pubsub/rabbitmq/rabbitmq_test.go
Comment (berndverst) [inline tests/certification/pubsub/rabbitmq/rabbitmq_test.go:1]: More is needed here -- please revert this change here.

If possible I'd like to see something that actually **additionally** shows either quorum or stream queues in use. Do you think that is possible?
If so create an entirely separate new `flow` run block. And if necessary you may need to define a different docker compose file -- depending on whether RabbitMQ needs to be initialized with different settings for different queue types.

if there is a behavior change between classic queues and one of the other queue types which we can assert -- that would be ideal!

**Review by berndverst — CHANGES_REQUESTED** [review]

File: pubsub/rabbitmq/rabbitmq.go
Comment (aaguilartablada) [reply pubsub/rabbitmq/rabbitmq.go:410]: I will change it to indicate queue type on req level instead of component

File: pubsub/rabbitmq/rabbitmq.go
Comment (aaguilartablada) [reply pubsub/rabbitmq/rabbitmq.go:446]: You are right. My fault. As it only applies to stream queues, I'm going to change it to be defined in req metadata

File: tests/certification/pubsub/rabbitmq/rabbitmq_test.go
Comment (aaguilartablada) [reply tests/certification/pubsub/rabbitmq/rabbitmq_test.go:1]: Ok! I will work in it

File: pubsub/rabbitmq/rabbitmq.go
Comment (DeepanshuA) [inline pubsub/rabbitmq/rabbitmq.go:51]: Should we rename parameter name to reflect that it is for stream specifically? Something like, StreamMaxAge.

Just thinking, if we should use directly amqp constant here as well? 
https://github.com/rabbitmq/amqp091-go/blob/09d223a6803ba6824f3c311c82a3f630c1b162a9/types.go#L255

**Comment by aaguilartablada** [comment]
Hi again @DeepanshuA. 

Finally I discarded stream queues as I've never worked with them and their functionality is radically different from classic and quorum queues. Furthermore, it requires another package (github.com/rabbitmq/rabbitmq-stream-go-client/pkg/amqp) and several code changes to work. Maybe another contributor has experience whit this kind of queues and can create a dedicated PR.

The difference between classic and quorum queues is in the server and how it replicates and manage data in memory and disk. For this reason I didn't create a new flow for quorum queues. Instead, I made a change in an existing flow in which we have two consumers. One consumer creates quorum queues and the other creates classic queues. What do you think?

**Comment by berndverst** [comment]
Unfortunately this will not make it into release 1.11 at this point.

**Review by aaguilartablada — COMMENTED** [review]
can you review again, please?

**Review by aaguilartablada — COMMENTED** [review]
@berndverst I don't know why the PR is stuck pending on requested changes that I already resolved.

File: pubsub/rabbitmq/metadata.go
Comment (aaguilartablada) [reply pubsub/rabbitmq/metadata.go]: deleted

File: pubsub/rabbitmq/metadata.go
Comment (aaguilartablada) [reply pubsub/rabbitmq/metadata.go]: deleted

File: pubsub/rabbitmq/rabbitmq.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq.go:416]:
```suggestion
			return nil, fmt.Errorf("invalid queue type %s. Valid types are %s and %s", val, amqp.QueueTypeClassic, amqp.QueueTypeQuorum)
```

File: pubsub/rabbitmq/rabbitmq.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq.go:346]:
```suggestion
	case failed := <-ackCh:
		if failed {
```

File: pubsub/rabbitmq/rabbitmq.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq.go:414]:
```suggestion
	if val := req.Metadata[reqMetadataQueueTypeKey]; val != "" {
```

File: pubsub/rabbitmq/rabbitmq.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq.go:420]:
```suggestion
	} else {
```

File: pubsub/rabbitmq/rabbitmq_test.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq_test.go:103]:
```suggestion
	assert.NoError(t, err)
```

File: pubsub/rabbitmq/rabbitmq_test.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq_test.go:107]:
```suggestion
	assert.NoError(t, err)
```

File: pubsub/rabbitmq/rabbitmq_test.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq_test.go:118]:
```suggestion
	assert.NoError(t, err)
```

File: pubsub/rabbitmq/rabbitmq_test.go
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq_test.go:125]:
```suggestion
	assert.Error(t, err)
```

**Comment by ItalyPaleAle** [comment]
@aaguilartablada will merge when tests pass

Can you please open an issue in dapr/docs to get the docs for this component updated?

**Review by ItalyPaleAle — APPROVED** [review]

**Pull request merged**
